### PR TITLE
Support maps in `BuildProperties`

### PR DESCRIPTION
**Pull request by mr-archano**
> Addresses #18 

### Scope of the PR
At the moment the plugin introduces a set of build properties backed by system environment variables, eg:
```
buildProperties.env['JAVA_HOME']
```

Altho this has been proved to be useful, I would prefer to let the user decide whether to define an entry in the `buildProperties` container and just expose a simple way of creating an instance of `Entries` from `System.getnv()` (or a map in general!).

### Considerations/Implementation Details

- Introduced `MapEntries` as implementation of `Entries` backed by a `Map<String, Object>`
- Exposed way to fill a `BuildProperties` instance using `MapEntries`
- Removed `EnvironmentPropertiesEntries` and built-in `buildProperties.env` because not needed anymore.

### CODE REVIEW

File: plugin/src/test/groovy/com/novoda/buildproperties/test/ExtendedTruth.groovy [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import com.google.common.collect.*
+import com.google.common.truth.*
+import com.google.common.util.concurrent.AtomicLongMap
+import com.novoda.buildproperties.CompositeException
+import com.novoda.buildproperties.Entry
+
+import javax.annotation.Nullable
+
+/**
+ * This utility class serves as static extension of {@link Truth},
```
Comment (tasomaniac) [inline plugin/src/test/groovy/com/novoda/buildproperties/test/ExtendedTruth.groovy]: Can you add these 2 extra static methods as a Groovy extension method on `Truth`? Then you don't need to implement all the others. 
I found this as a sample, the example at the end adds static method to `String` class.

File: sample/app/build.gradle [hunk 2]
```diff
@@ -25,6 +25,9 @@ buildProperties {
     releaseSigning {
         file rootProject.file('properties/releaseSigning.properties')
     }
+    env {
```
Comment (tasomaniac) [inline sample/app/build.gradle]: I think not just in the sample, we should also update `README` to include this change.

File: plugin/src/test/groovy/com/novoda/buildproperties/EnvironmentPropertiesEntriesTest.groovy [hunk 3]
```diff
@@ -1,70 +0,0 @@
...
-import org.gradle.testfixtures.ProjectBuilder
-import org.junit.Before
-import org.junit.Rule
-import org.junit.Test
-import org.junit.contrib.java.lang.system.EnvironmentVariables
-import org.junit.rules.TemporaryFolder
-
-import static com.google.common.truth.Truth.assertThat
-
-class EnvironmentPropertiesEntriesTest {
```
Comment (tasomaniac) [inline plugin/src/test/groovy/com/novoda/buildproperties/EnvironmentPropertiesEntriesTest.groovy]: Just to understand. I guess the integration test still tests, setting the real EnvironmentVariable thanks to the change in the sample? 💯

File: sample/app/build.gradle (hunk 2 above)
Comment (mr-archano) [inline sample/app/build.gradle]: I totally forgot, you're right 👍

File: plugin/src/test/groovy/com/novoda/buildproperties/EnvironmentPropertiesEntriesTest.groovy (hunk 3 above)
Comment (mr-archano) [inline plugin/src/test/groovy/com/novoda/buildproperties/EnvironmentPropertiesEntriesTest.groovy]: Yep, the integration test uses environment variables still.

File: plugin/src/test/groovy/com/novoda/buildproperties/test/ExtendedTruth.groovy (hunk 1 above)
Comment (tasomaniac) [inline plugin/src/test/groovy/com/novoda/buildproperties/test/ExtendedTruth.groovy]: Here is the link :) http://mrhaki.blogspot.de/2013/01/groovy-goodness-adding-extra-methods.html

Comment (mr-archano) [inline plugin/src/test/groovy/com/novoda/buildproperties/test/ExtendedTruth.groovy]: @tasomaniac I tried to make this work but no luck. Happy to keep this as it is now?

Comment (tasomaniac) [inline plugin/src/test/groovy/com/novoda/buildproperties/test/ExtendedTruth.groovy]: Of course. 👍 Was just an optional idea.

File: plugin/src/test/groovy/com/novoda/buildproperties/MapEntriesTest.groovy [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Test
+    public void shouldContainDefinedKey() {
+        def entries = new MapEntries([foo: 'bar'])
+
+        assertThat(entries.contains('foo')).isTrue()
+    }
+
+    @Test
+    public void shouldProvidValueForDefinedKey() {
```
Comment (ataulm) [inline plugin/src/test/groovy/com/novoda/buildproperties/MapEntriesTest.groovy]: provide (typo)

File: plugin/src/test/groovy/com/novoda/buildproperties/MapEntriesTest.groovy [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        assertThat(entry).hasValue('bar')
+    }
+
+    @Test
+    public void shouldSupportKeysListing() {
+        def entries = new MapEntries([foo: 'bar', x: 'y'])
+
+        Enumeration<String> keys = entries.keys
+
+        assertThat(Collections.list(keys)).containsAllOf('foo', 'x').inOrder()
```
Comment (ataulm) [inline plugin/src/test/groovy/com/novoda/buildproperties/MapEntriesTest.groovy]: is the `inOrder` important? i.e. is it a contract of maps?

Comment (mr-archano) [inline plugin/src/test/groovy/com/novoda/buildproperties/MapEntriesTest.groovy]: Good point. Checking the order is unnecessary 👍

**Pull request merged**